File: python/insert_interval.py
class Solution:
#     # suboptimal sorting approach
#     def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
#         # merge and sort new interval into existing ones
#         intervals.append(newInterval)
#         intervals = sorted(intervals, key = lambda interval: interval[0])

#         mergedIntervals = [intervals[0]]

#         for i in range(1, len(intervals)):
#             numMerged = len(mergedIntervals)

#             # current interval is beyond previously written, add it
#             if intervals[i][0] > mergedIntervals[numMerged - 1][1]:
#                 mergedIntervals.append(intervals[i])
#             # current interval starting value "dips" into previously written
#             # since intervals are sorted by starting value), extend if possible
#             else:
#                 mergedIntervals[numMerged - 1][1] = max(intervals[i][1], mergedIntervals[numMerged - 1][1])

#         return mergedIntervals

    # optimal greedy approach
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        mergedIntervals = []

        for i in range(len(intervals)):
            # new interval falls before current one, add it and remaining intervals
            if newInterval[1] < intervals[i][0]:
                mergedIntervals.append(newInterval)
                mergedIntervals.extend(intervals[i:])
                return mergedIntervals
            # new interval is beyond current one, add current one and continue
            if newInterval[0] > intervals[i][1]:
                mergedIntervals.append(intervals[i])
            # new interval overlaps with current interval, update new interval
            # as local max of overlapping start and end bounds
            else:
                newInterval[0] = min(newInterval[0], intervals[i][0])
                newInterval[1] = max(newInterval[1], intervals[i][1])

        # new interval (or local max merged interval) is beyond all intervals
        mergedIntervals.append(newInterval)
        return mergedIntervals

File: cpp/insert_interval.cpp
class Solution {
public:
    // test cases:
    //                                      i
    // input:  [[1,2],[3,5],[6,7],[8,10],[12,16]] newInterval = [4,8],[3,10]
    // output: [[1,2],[3,10],[12,16]]
    //                            i
    // input:  [[2,3],[4,5],[6,7]] newInterval = [1,8]
    // output: [[1,8]]
    //                      i
    // input:  [[1,1],[2,2]] newInterval = [2,2]
    // output: [[1,1],[2,2]]
    //                i
    // input:  [[1,1]] newInterval = [2,2]
    // output: [[1,1],[2,2]]
    //                i
    // input:  [[1,1]] newInterval = [0,2]
    // output: [[0,2]]
    //            i
    // input:  [[1,1]] newInterval = [0,0]
    // output: [[0,0][1,1]]
    //          i
    // input:  [] newInterval = [0,0]
    // output: [[0,0]]
    //
    // optimal greedy approach
    vector<vector<int>> insert(vector<vector<int>> &intervals, vector<int> &newInterval) {
        vector<vector<int>> mergedIntervals;

        for (int i = 0; i < intervals.size(); ++i) {
            // new interval falls before current one, add it and remaining intervals
            if (newInterval[1] < intervals[i][0]) {
                mergedIntervals.push_back(newInterval);
                vector<vector<int>> rest = { intervals.begin() + i, intervals.end() };
                mergedIntervals.insert(mergedIntervals.end(), rest.begin(), rest.end());
                return mergedIntervals;
            }
            // new interval is beyond current one, add current one and continue
            else if (intervals[i][1] < newInterval[0]) {
                mergedIntervals.push_back(intervals[i]);
            }
            // new interval overlaps with current interval, expand new interval
            // by taking max of overlapping start and end bounds (could absorb)
            else {
                newInterval[0] = min(intervals[i][0], newInterval[0]);
                newInterval[1] = max(intervals[i][1], newInterval[1]);
            }
        }

        // new interval is either beyond all other intervals, or it expanded and
        // absorbed all other intervals all the way to the end
        mergedIntervals.push_back(newInterval);
        return mergedIntervals;
    }
};
